allennlp/models/bert_for_classification.py

```diff
@@ -0,0 +1,141 @@
+    label_namespace : ``str``, optional (default : "labels")
+        Used to determine the number of classes if ``num_labels`` is not supplied.
+    trainable : ``bool``, optional (default : True)
+        If True, the weights of the pretrained BERT model will be updated during training.
+        Otherwise, they will be frozen and only the final linear layer will be trained.
+    initializer : ``InitializerApplicator``, optional
+        If provided, will be used to initialize the final linear layer *only*.
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 bert_model: Union[str, BertModel],
+                 num_labels: int = None,
+                 index: str = "bert",
+                 label_namespace: str = "labels",
+                 trainable: bool = True,
+                 initializer: InitializerApplicator = InitializerApplicator()) -> None:
+        super().__init__(vocab)
+
+        if isinstance(bert_model, str):
+            self.bert_model = BertModel.from_pretrained(bert_model)
+        else:
+            self.bert_model = bert_model
+
+        self.bert_model.requires_grad = trainable
+
+        in_features = self.bert_model.config.hidden_size
+
+        if num_labels:
+            out_features = num_labels
+        else:
+            out_features = vocab.get_vocab_size(label_namespace)
+
+        self._classification_layer = torch.nn.Linear(in_features, out_features)
+        self._accuracy = CategoricalAccuracy()
+        self._loss = torch.nn.CrossEntropyLoss()
+        self._index = index
+        initializer(self._classification_layer)
+
+    def forward(self,  # type: ignore
+                tokens: Dict[str, torch.LongTensor],
+                label: torch.IntTensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Parameters
+        ----------
+        tokens : Dict[str, torch.LongTensor]
+            From a ``TextField`` (that has a bert-pretrained token indexer)
+        label : torch.IntTensor, optional (default = None)
+            From a ``LabelField``
+
+        Returns
+        -------
+        An output dictionary consisting of:
+
+        logits : torch.FloatTensor
+            A tensor of shape ``(batch_size, num_labels)`` representing
+            unnormalized log probabilities of the label.
+        probs : torch.FloatTensor
+            A tensor of shape ``(batch_size, num_labels)`` representing
+            probabilities of the label.
+        loss : torch.FloatTensor, optional
+            A scalar loss to be optimised.
+        """
+        input_ids = tokens[self._index]
+        token_type_ids = tokens[f"{self._index}-type-ids"]
+        input_mask = (input_ids != 0).long()
+
+        _, pooled = self.bert_model(input_ids=input_ids,
+                                    token_type_ids=token_type_ids,
+                                    attention_mask=input_mask)
+
+        # apply classification layer
```

Comment:
It's customary to add dropout here, not sure how important that is: https://github.com/huggingface/pytorch-pretrained-BERT/blob/3ae8c8be1e3fc770968cd3fdb3b643e0b166e540/pytorch_pretrained_bert/modeling.py#L1058

Reply:
done, thanks